lib/auth/webauthn/messages_test.go

```diff
@@ -56,3 +57,270 @@ func TestCredentialAssertionResponse_json(t *testing.T) {
+		wantErr   string
+	}{
+		{
+			name:      "ok", // check that good params are good
+			assertion: okAssertion,
+			wantErr:   "",
+		},
+		{
+			name:    "nil assertion",
+			wantErr: "assertion required",
+		},
+		{
+			name:      "assertion without challenge",
+			assertion: &nilChallengeAssertion,
+			wantErr:   "challenge",
+		},
+		{
+			name:      "assertion without RPID",
+			assertion: &emptyRPIDAssertion,
+			wantErr:   "relying party ID",
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			err := test.assertion.Validate()
+			if test.wantErr != "" {
+				require.ErrorContains(t, err, test.wantErr, "Validate returned err = %q, want %q", err, test.wantErr)
+			} else {
+				require.NoError(t, err, "Validate returned err %v, want nil", err)
+			}
+		})
+	}
+}
+
+func TestRequireResidentKey(t *testing.T) {
+	tests := []struct {
+		name    string
+		in      protocol.AuthenticatorSelection
+		want    bool
+		wantErr string
+	}{
+		{
+			name: "nothing set",
+			in:   protocol.AuthenticatorSelection{},
+			want: false,
+		},
+		{
+			name: "mismatch of ResidentKey and RequireResidentKey - return error",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementDiscouraged,
+				RequireResidentKey: protocol.ResidentKeyRequired(),
+			},
+			wantErr: "invalid combinantion of ResidentKey: discouraged and RequireResidentKey: true",
+		},
+		{
+			name: "other mismatch of ResidentKey and RequireResidentKey - return error",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementRequired,
+				RequireResidentKey: protocol.ResidentKeyUnrequired(),
+			},
+			wantErr: "invalid combinantion of ResidentKey: required and RequireResidentKey: false",
+		},
+		{
+			name: "support nil RequireResidentKey",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        "",
+				RequireResidentKey: nil,
+			},
+			want: false,
+		},
+		{
+			name: "ResidentKey preferred result in false",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        protocol.ResidentKeyRequirementPreferred,
+				RequireResidentKey: nil,
+			},
+			want: false,
+		},
+		{
+			name: "ResidentKey required",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey: protocol.ResidentKeyRequirementRequired,
+			},
+			want: true,
+		},
+		{
+			name: "ResidentKey discouraged",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey: protocol.ResidentKeyRequirementDiscouraged,
+			},
+			want: false,
+		},
+		{
+			name: "use RequireResidentKey required if ResidentKey empty",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        "",
+				RequireResidentKey: protocol.ResidentKeyRequired(),
+			},
+			want: true,
+		},
+		{
+			name: "use RequireResidentKey unrequired if ResidentKey empty",
+			in: protocol.AuthenticatorSelection{
+				ResidentKey:        "",
+				RequireResidentKey: protocol.ResidentKeyUnrequired(),
+			},
+			want: false,
+		},
```

Comment:
```suggestion
		{
			name: "use RequireResidentKey unrequired if ResidentKey empty",
			in: protocol.AuthenticatorSelection{
				ResidentKey:        "",
				RequireResidentKey: protocol.ResidentKeyRequired(),
			},
			want: true,
		},
```

Otherwise it can't be distinguished from the ""/nil scenario.

Reply:
we already have test for required above. I will remove it.